src/main/java/com/github/firmwehr/gentle/firm/optimization/EscapeAnalysisOptimization.java

```diff
@@ -0,0 +1,392 @@
+	}
+
+	private static LocalAddress fromPred(Node pred) {
+		return fromPredIfMatches(pred).orElseThrow(
+			() -> new InternalCompilerException("Don't know how to transform " + pred + " into local address"));
+	}
+
+	private static Optional<LocalAddress> fromPredIfMatches(Node pred) {
+		if (pred instanceof Proj proj) {
+			return Optional.of(new LocalAddress(proj, 0)); // 0th field in class
+		} else if (pred instanceof Add add) {
+			if (add.getLeft() instanceof Proj proj && add.getRight() instanceof Const val) {
+				return Optional.of(new LocalAddress(proj, val.getTarval().asLong()));
+			} else if (add.getRight() instanceof Proj proj && add.getLeft() instanceof Const val) {
+				return Optional.of(new LocalAddress(proj, val.getTarval().asLong()));
+			}
+		}
+		return Optional.empty();
+	}
+
+	private static void walkDown(Node node, NodeVisitor visitor) {
+		node.accept(visitor);
+		node.markVisited();
+		if (node instanceof Load || node instanceof Store) {
+			// we don't need to follow Loads and Stores as their successors
+			// don't use the call result (and if they do, they are visited separately)
+			return;
+		}
+		for (BackEdges.Edge out : BackEdges.getOuts(node)) {
+			if ((out.node.getMode().equals(Mode.getT()) || out.node.getMode().equals(Mode.getP())) &&
+				!out.node.visited()) {
+				LOGGER.debug("visiting %s", out.node);
+				walkDown(out.node, visitor);
+			} else {
+				LOGGER.debug("not visiting %s (visited = %s)", out.node, out.node.visited());
+			}
+		}
+	}
+
+	// return calls that do not escape
+	private Optional<Call> filterNonEscapingAllocations(Call allocationCall) {
+		if (BackEdges.getNOuts(allocationCall) < 2) {
+			return Optional.of(allocationCall);
+		}
+		Node projResT = successor(allocationCall, false);
+		Proj resultProj = (Proj) successor(projResT, false);
+		if (!escapes(resultProj)) {
+			return Optional.of(allocationCall);
+		}
+		return Optional.empty();
+	}
+
+	private boolean escapes(Proj callResProj) {
+		Deque<Node> deepOuts = new ArrayDeque<>();
+		deepOuts.add(callResProj);
+		while (!deepOuts.isEmpty()) {
+			Node next = deepOuts.removeLast();
+			for (BackEdges.Edge edge : BackEdges.getOuts(next)) {
+				if (edge.node.getMode().equals(Mode.getM())) {
+					LOGGER.debug("stop at %s (memory)", edge.node);
+					continue;
+				}
+				if (edge.node instanceof Call || edge.node instanceof Return) {
+					LOGGER.debug("%s escapes on %s", callResProj, edge.node);
+					return true;
+				}
+				if (edge.node instanceof Load load) {
+					if (fromPredIfMatches(load.getPtr()).isPresent()) {
+						// Load from this call => ignore (would be replaced)
+						LOGGER.debug("stop at %s", edge.node);
+						continue;
+					}
+					LOGGER.debug("%s escapes on %s (invalid address)", callResProj, edge.node);
+					return true;
+				}
+				if (edge.node instanceof Store store) {
+					if (!callResProjIsReachableFromStoreValue(store.getValue(), callResProj, new HashSet<>())) {
```

Comment:
What does this check do? Is the comment block below explaining whats going on here? Probably should be move up